disk/disk_windows.go

```diff
@@ -81,84 +93,86 @@ func UsageWithContext(_ context.Context, path string) (*UsageStat, error) {
+// It uses procGetLogicalDriveStringsW to get drives with drive letters and procFindFirstVolumeW to get volumes without drive letters.
+// Since the api calls don't have a timeout, this method uses context to set deadline by users.
 func PartitionsWithContext(ctx context.Context, _ bool) ([]PartitionStat, error) {
-	warnings := Warnings{
-		Verbose: true,
-	}
-
-	var errLogicalDrives error
+	warnings := Warnings{Verbose: true}
+	var errInitialCall error
 	retChan := make(chan PartitionStat)
 	quitChan := make(chan struct{})
 	defer close(quitChan)
+	processedPaths := make(map[string]struct{})
 
 	getPartitions := func() {
 		defer close(retChan)
 
+		// Get drives with drive letters (including remote drives, ex: SMB shares)
 		lpBuffer := make([]byte, 254)
-
-		diskret, _, err := procGetLogicalDriveStringsW.Call(
+		if diskret, _, err := procGetLogicalDriveStringsW.Call(
 			uintptr(len(lpBuffer)),
-			uintptr(unsafe.Pointer(&lpBuffer[0])))
-		if diskret == 0 {
-			errLogicalDrives = err
+			uintptr(unsafe.Pointer(&lpBuffer[0]))); diskret == 0 {
+			errInitialCall = err
 			return
 		}
 		for _, v := range lpBuffer {
 			if v >= 65 && v <= 90 {
 				path := string(v) + ":"
-				typepath, _ := windows.UTF16PtrFromString(path)
-				typeret := windows.GetDriveType(typepath)
-				switch typeret {
-				case windows.DRIVE_UNKNOWN:
-					continue
-				case windows.DRIVE_REMOVABLE,
-					windows.DRIVE_FIXED,
-					windows.DRIVE_REMOTE,
-					windows.DRIVE_CDROM:
-					lpVolumeNameBuffer := make([]byte, 256)
-					lpVolumeSerialNumber := int64(0)
-					lpMaximumComponentLength := int64(0)
-					lpFileSystemFlags := int64(0)
-					lpFileSystemNameBuffer := make([]byte, 256)
-					volpath, _ := windows.UTF16PtrFromString(string(v) + ":/")
-					driveret, _, err := procGetVolumeInformation.Call(
-						uintptr(unsafe.Pointer(volpath)),
-						uintptr(unsafe.Pointer(&lpVolumeNameBuffer[0])),
-						uintptr(len(lpVolumeNameBuffer)),
-						uintptr(unsafe.Pointer(&lpVolumeSerialNumber)),
-						uintptr(unsafe.Pointer(&lpMaximumComponentLength)),
-						uintptr(unsafe.Pointer(&lpFileSystemFlags)),
-						uintptr(unsafe.Pointer(&lpFileSystemNameBuffer[0])),
-						uintptr(len(lpFileSystemNameBuffer)))
-					if driveret == 0 {
-						switch typeret {
-						case windows.DRIVE_REMOVABLE, windows.DRIVE_REMOTE, windows.DRIVE_CDROM:
-							continue // device is not ready will happen if there is no disk in the drive
-						}
-						warnings.Add(err)
-						continue
-					}
-					opts := []string{"rw"}
-					if lpFileSystemFlags&fileReadOnlyVolume != 0 {
-						opts = []string{"ro"}
-					}
-					if lpFileSystemFlags&fileFileCompression != 0 {
-						opts = append(opts, "compress")
+				if partitionStat, warning := buildPartitionStat(path); warning == nil {
+					processedPaths[partitionStat.Mountpoint+"\\"] = struct{}{}
+					select {
+					case retChan <- partitionStat:
+					case <-quitChan:
+						return
```

Comment:
Could you add `<-ctx.Done()` in this `select` to stop this goroutine when cancelled?